Add more docs on durable state (#30403)
akka/akka: akka-docs/src/main/paradox/persistence-query-durable-state.md

---
project.description: Query side to Akka Persistence allowing for building CQRS applications using durable state.
---
# Persistence Query

## Dependency

To use Persistence Query, you must add the following dependency in your project:

@@dependency[sbt,Maven,Gradle] {
  bomGroup=com.typesafe.akka bomArtifact=akka-bom_$scala.binary.version$ bomVersionSymbols=AkkaVersion
  symbol1=AkkaVersion
  value1="$akka.version$"
  group=com.typesafe.akka
  artifact=akka-persistence-query_$scala.binary.version$
  version=AkkaVersion
}

This will also add dependency on the @ref[Akka Persistence](persistence.md) module.

## Introduction

> Akka persistence query provides a query interface to @ref:[Durable State Behaviors](typed/persistence-durable-state.md).
> These queries are based on asynchronous streams, very much like the ones offered in the @ref:[Event Sourcing](persistence-query.md)
> based implementation. Various state store plugins can implement these interfaces in order to expose their query capabilities.

Reviewer: ```suggestion
Akka persistence query provides a query interface to @ref:[Durable State Behaviors](typed/persistence-durable-state.md).
These queries are based on asynchronous streams. These streams are similar to the ones offered in the @ref:[Event Sourcing](persistence-query.md)
based implementation. Various state store plugins can implement these interfaces to expose their query capabilities.
```

Author: done ..

> One of the rationales behind having a separate query module for Akka Persistence is for implementing the so-called 
> query side (also known as "read side") in the popular CQRS architecture pattern - in which the writing side of the 
> application (e.g. implemented using Akka persistence) is completely separated from the "query side".

Reviewer: ```suggestion
A rationale for having a separate query module, for Akka Persistence, is to implement the so-called 
query side, or read side, in the popular CQRS architecture pattern. In this pattern, the writing side of the application, implemented using Akka Persistence, is completely separated from the query side.
```

Author: done ..

## Using query with Akka Projections

Akka Persistence and Akka Projections together can be used to develop a CQRS application. In the application the 
durable state is stored in a database and fetched as an asynchronous stream to the user. Currently queries on 

> durable state, provided by the interface in `DurableStateStoreQuery`, is used to implement tag based searches in 

Reviewer 2: It is not clear how to get this interface, and use it. Maybe we should show something like this:

```
    val durableStateStoreQuery =
      DurableStateStoreRegistry(system)
        .getDurableStateStoreFor[DurableStateStoreQuery[A]](classOf[DurableStateStoreQuery[A]], pluginId)
```

Reviewer 2: (when akka projection doc is ready we should point them to that and probably explain that it is better to use that instead of the more 'internal' `DurableStateStoreQuery`)

> Akka Persistence and Akka Projections together can be used to develop a CQRS application where the 
> durable state is stored in a database and fetched as an asynchronous stream to the user. Currently queries on 
> durable state, provided by the interface in `DurableStateStoreQuery`, is used to implement Akka Projections' 
> tag based searches.

Reviewer: ```suggestion
Akka Persistence and Akka Projections together can be used to develop a CQRS application. In this application, the
durable state is stored in a database and fetched as an asynchronous stream to the user. Currently, queries on durable state, provided by the interface in `DurableStateStoreQuery`, are used to implement tag-based searches for Akka Projections.
.
```

Author: done ..
